pkg/endpoints/endpoints.go

```diff
@@ -70,28 +97,63 @@ func (e *endpointGroup) getPort(portName string) int32 {
 }
 
 func (g *endpointGroup) lenIPs() int {
-	g.mtx.Lock()
-	defer g.mtx.Unlock()
+	g.mtx.RLock()
+	defer g.mtx.RUnlock()
 	return len(g.endpoints)
 }
 
 func (g *endpointGroup) setIPs(ips map[string]struct{}, ports map[string]int32) {
 	g.mtx.Lock()
-	defer g.mtx.Unlock()
-
 	g.ports = ports
 	for ip := range ips {
 		if _, ok := g.endpoints[ip]; !ok {
-			g.endpoints[ip] = endpoint{inFlight: &atomic.Int64{}}
+			g.endpoints[ip] = endpoint{inFlight: &atomic.Int64{}, terminated: make(chan struct{})}
 		}
 	}
-	for ip := range g.endpoints {
+	for ip, endpoint := range g.endpoints {
 		if _, ok := ips[ip]; !ok {
 			delete(g.endpoints, ip)
+			close(endpoint.terminated)
 		}
 	}
+	g.mtx.Unlock()
 
-	if len(g.endpoints) > 0 {
-		g.active.Broadcast()
+	// notify waiting requests
+	if len(ips) > 0 {
+		g.broadcastEndpoints()
 	}
 }
+
+func (g *endpointGroup) broadcastEndpoints() {
+	g.bmtx.Lock()
+	defer g.bmtx.Unlock()
+
+	close(g.bcast)
+	g.bcast = make(chan struct{})
+}
+
+func (e *endpointGroup) AddInflight(addr string, cancelRequest context.CancelFunc) (func(), error) {
```

Comment:
Not sure if it makes sense to actively cancel any inflight request here. We could also rely on the k8s internals to handle it.

Reply:
I would avoid actively cancelled to avoid complexity